daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala

```diff
@@ -41,13 +41,22 @@ import com.digitalasset.daml.lf.speedy.{Command => SpeedyCommand}
   * even if `result1` and `result2` request to dereference the same id.
   * <p>
   *
-  * This class is thread safe.
+  * The class requires a pseudo random generator (`nextRandomInt`) to randomize the
+  * submission time. This generator does not have to be cryptographically secure.
+  * <p>
+  *
+  * This class is thread safe as long `nextRandomInt` is.
   */
-final class Engine {
+final class Engine(nextRandomInt: () => Int) {
   private[this] val _compiledPackages: MutableCompiledPackages = ConcurrentCompiledPackages()
   private[this] val _commandTranslation: CommandPreprocessor = new CommandPreprocessor(
     _compiledPackages)
 
+  // We want to make the submissionTime different from the ledgerTime,
```

Comment:
A comment explaining why we want to make it different would be helpful here.

Reply:
done